megamek/src/megamek/common/Hex.java

```diff
@@ -631,82 +631,102 @@ public boolean isClearHex() {
-        boolean rv = true;
-
+        boolean valid = true;
+        
+        // When no StringBuffer is passed, use a dummy
+        // to avoid numerous null checks
+        if (errBuff == null) {
+            errBuff = new StringBuffer();
+        }
+        
         // Check individual terrains for validity
         for (ITerrain terrain : terrains.values()) {
             if (terrain == null) {
+                valid = false;
+                errBuff.append("Hex contains a null terrain!\n");
                 continue;
             }
-            StringBuffer currBuff = new StringBuffer();
-            boolean isValid = terrain.isValid(currBuff);
-            if (!isValid && (errBuff == null)) {
-                return false;
-            } else if (!isValid) {
-                rv = false;
+            StringBuffer terrainErr = new StringBuffer();
+            if (!terrain.isValid(terrainErr)) {
+                valid = false;
                 if (errBuff.length() > 0) {
                     errBuff.append("\n");
                 }
-                errBuff.append(currBuff);
+                errBuff.append(terrainErr);
             }
         }
-        // Some terrains must be grouped, check for those.
 
         // Rapids
         if ((containsTerrain(Terrains.RAPIDS))) {
             if (!containsTerrain(Terrains.WATER)) {
-                rv = false;
-                if (errBuff != null) {
-                    errBuff.append("Rapids must occurr within water!\n");
-                }
+                valid = false;
+                errBuff.append("Rapids must occur within water!\n");
             }
             if (this.depth() <1) {
-                rv = false;
-                if (errBuff != null) {
-                    errBuff.append("Rapids must occurr in depth 1 or greater!\n");
-                }
+                valid = false;
+                errBuff.append("Rapids must occurr in depth 1 or greater!\n");
             }
         }
 
+        // Foliage (Woods and Jungles)
+        if (containsTerrain(Terrains.WOODS) && containsTerrain(Terrains.JUNGLE)) {
+            valid = false;
+            errBuff.append("Woods and Jungle cannot appear in the same hex!\n");
+        }
+        if ((containsTerrain(Terrains.WOODS) || containsTerrain(Terrains.JUNGLE))
+                && containsTerrain(Terrains.FOLIAGE_ELEV)) {
+            int wl = terrainLevel(Terrains.WOODS);
+            int jl = terrainLevel(Terrains.JUNGLE);
+            int el = terrainLevel(Terrains.FOLIAGE_ELEV);
+            
+            boolean isLightOrHeavy = wl == 1 || jl == 1 || wl == 2 || jl == 2;
+            boolean isUltra = wl == 3 || jl == 3;
+            
+            if (! ((el == 1) || (isLightOrHeavy && el == 2) || (isUltra && el == 3))) {
+                valid = false;
+                errBuff.append("Foliage elevation is wrong, must be 1 or 2 for Light/Heavy and 1 or 3 for Ultra Woods/Jungle!\n");
```

Comment:
I'm confused by this.

Reply:
The elevation can always be 1 (regardless of light, heavy or ultra), it can also be 2 but only in light or heavy woods/jungle and in ultra woods/jungle 3 is an allowed value. Not quite sure how to answer ...

Reply:
Ok, I just didn't realize that ultra terrain is either 1 or 3. Kind of weird, but it is what it is.